### PR TITLE
[Site Editor]: Add create pattern button in patterns page

### LINKED ISSUE #58198: Make 'Add' buttons consistent across data views, simplify labels
add

- [ ] "Add **new** $post_type" is unnecessarily verbose. We can shorten this to "Add $post_type". Trac ticket: https://core.trac.wordpress.org/ticket/61219#ticket
- [x] The add flow for Patterns should match other data views. IE a primary button in the page header. For now this button will open the same menu as the current `+` in the sidebar. The `+` button can be removed.

### LINKED ISSUE #59099: Site Editor > Patterns: Consider adding a "Add new pattern" button
In watching some folks test the new Data Views for 6.5, one of the participants mentioned how hard it is to figure out how to create your own patterns with the + button being so easy to miss. When compared to the other sections, it stands out even more as the Patterns section shows all patterns but doesn't have that ever present "Add new ____" button clearly placed. This feels odd considering patterns are likely something someone will build more than new templates. We want folks to use these and they aren't meant to be overly advanced! Here's a snapshot:

**Site Editor > Templates > Manage all templates:**

Screenshot 2024-02-15 at 9 42 17 AM

**Site Editor > Patterns**
Screenshot 2024-02-15 at 9 42 33 AM

**Site Editor > Patterns > Manage all template parts**

Screenshot 2024-02-15 at 9 51 03 AM

**Site Editor > Pages > Manage all pages**
Screenshot 2024-02-15 at 9 42 06 AM

## How can we improve this?

My brain goes in a few directions:

- Add a "Add new pattern" to each section of patterns. We'd have to decide if you're in the "Banner" category if clicking "Add new" would automatically add that as a category or if each were distinct ("my patterns" would have to be no category since it isn't tied to one).
- Add "Add a new pattern" after someone adds an initial pattern under just the "my patterns" section and "all patterns" (or just one section). This would at least tie the action to creating your own pattern but doesn't seem great nor consistent. I could see it feeling like something you're stumbling on and it stands in contrast to what we do in other sections.

Something else entirely :)  @WordPress/gutenberg-design as I imagine this was an intentional decision.

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Resolves: https://github.com/WordPress/gutenberg/issues/59099
Resolves: https://github.com/WordPress/gutenberg/issues/58198

This PR adds a `Create pattern` button in patterns page header and uses the same modal for creating patterns/template parts in the sidebar, as suggested here: https://github.com/WordPress/gutenberg/issues/58198

## Testing Instructions
1. Go to patterns page and observe the new `Create pattern` button
2. Test the menu items that work properly

## Screenshots or screencast 

Screenshot 2024-05-20 at 7 44 58 PM

### CODE REVIEW

**Comment by annezazu** [comment]
So excited to see this being worked on! Minor but, for consistency's sake, can we update this to "Add new pattern"? For context:

Screenshot 2024-03-29 at 2 03 03 PM
Screenshot 2024-03-29 at 2 02 54 PM
Screenshot 2024-03-29 at 2 03 11 PM

In reading through https://github.com/WordPress/gutenberg/issues/58198, it seems like we're trying to move towards "Add new" which this doesn't do as well. As a result, for now, I'd argue for having it as "Add new pattern" and then updating all to "Add new" or whatever is decided upon in the future cc @jameskoster for a confidence check.

**Comment by richtabor** [comment]
It should probably read "Add pattern"—but I'm also not 100% about this big primary buttons to be honest. It's the same style as the save buttons, in nearly the same position; I think it makes the experience less predictable. 

And also, do we have a pattern of using a primary button as a popover container? Feels a bit unexpected, as every other primary button is a key action taken, not leading to a post-selection decision. 

Curious what @WordPress/gutenberg-design thinks as well.

**Comment by t-hamano** [comment]
Based on the discussion in #58198, it seems like "Add new pattern" makes more sense than "Create pattern" following the logic in the core. In that case, should the popover menu items also be unified with "Add new"?

create-pattern

As mentioned in ToDo #58198, I think all additional buttons should eventually be replaced with `secondary` variations. Also, since the current button size is 36px, we should also consider that it will be updated to a standard size of 24px/32px/40px in the future.

As for the button itself, it looks like its position is slightly shifted upwards compared to other buttons, so we may need to adjust the CSS.

button-alignment

**Comment by ntsekouras** [comment]
I updated the button's variant to `secondary` and the copy to `add new pattern`. Any more feedback is welcome to consolidate all these 'add new' buttons.

**Comment by jasmussen** [comment]
> It should probably read "Add pattern"—but I'm also not 100% about this big primary buttons to be honest. It's the same style as the save buttons, in nearly the same position; I think it makes the experience less predictable.

> And also, do we have a pattern of using a primary button as a popover container? Feels a bit unexpected, as every other primary button is a key action taken, not leading to a post-selection decision.

Agree with this. "Add Template Part" gets verbose too, probably in translations too, it doesn't feel the most scalable pattern.

**Comment by Mamaduka** [comment]
@jasmussen, I think we need to keep "what" in the button labels. See https://github.com/WordPress/gutenberg/issues/58198#issuecomment-1908302581 and https://github.com/WordPress/gutenberg/issues/58198#issuecomment-1957306529.

@ntsekouras, I think the label casing should match the other pages.

The UX is very different from other "Add New" buttons. Maybe instead of a dropdown, we should display a new pattern modal?

**Comment by jameskoster** [comment]
I don't have a strong opinion about the button labels, except that they should be consistent. "Add _x_" seems fine to me.

I agree the button opening a menu is a bit clunky, but I think that's a short-term thing until patterns and templates converge more comprehensively. If it's a big deal, then we could try a split-button, as we've been exploring for save + publish:

Add pattern

Eventually it should just open the 'Add pattern' UX (modal), and importing should probably be a secondary (or even tertiary) action. 

Speaking of secondary actions, if we keep the blue `primary` variant here then we can employ a pattern like this later which feels important:

Screenshot 2024-04-02 at 15 07 48

**Comment by alexstine** [comment]
Buttons/links must have enough context for users. Add New is not good enough for accessibility, it's why we changed all of them in Core. I am good with Add New X or Add X but not just Add or Add New.

https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html

This applies more to links over buttons but I think we should keep the same patterns for UI/UX consistency.

Thanks.

File: packages/edit-site/src/components/add-new-pattern/index.js [hunk 1]
```diff
@@ -103,10 +103,12 @@ export default function AddNewPattern() {
...
 				toggleProps={ {
-					as: SidebarButton,
+					as: showTextButton ? Button : SidebarButton,
+					variant: showTextButton ? 'secondary' : undefined,
 				} }
-				icon={ plus }
-				label={ __( 'Create pattern' ) }
+				icon={ showTextButton ? null : plus }
+				label={ showTextButton ? undefined : __( 'Add new pattern' ) }
+				text={ showTextButton ? __( 'Add new pattern' ) : undefined }
```
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-pattern/index.js]: Can we use the post type labels instead of hard coding custom labels?

**Comment by ntsekouras** [comment]
This should be ready for another round of reviews, based on the updated description of the [main issue](https://github.com/WordPress/gutenberg/issues/58198).

I can extract the labels update to remove the `new` for other post types too in a separate PR, to address https://core.trac.wordpress.org/ticket/61219#ticket.

File: lib/compat/wordpress-6.6/post.php [hunk 2]
```diff
@@ -33,7 +35,35 @@ function gutenberg_update_wp_template_labels( $labels ) {
...
+ *
+ * @param  object $labels Object with labels for the post type as member variables.
+ * @return object Object with all the labels as member variables.
+ */
+function gutenberg_update_page_labels( $labels ) {
+	$labels->add_new      = __( 'Add Page', 'gutenberg' );
+	$labels->add_new_item = __( 'Add Page', 'gutenberg' );
+	return $labels;
+}
+add_filter( 'post_type_labels_page', 'gutenberg_update_page_labels', 10, 1 );
```
Comment (youknowriad) [inline lib/compat/wordpress-6.6/post.php]: I think ideally we should make this change directly in Core because it will impact more than just the site editor and will not be limited to these post types. For instance I don't see why we should change these labels for "pages" but not for "posts". So I'd leave these label changes out of this PR personally.

File: packages/edit-site/src/components/create-template-part-modal/index.js [hunk 3]
```diff
@@ -39,7 +39,7 @@ import {
 } from '../../utils/template-part-create';
 
 export default function CreateTemplatePartModal( {
-	modalTitle = __( 'Create template part' ),
+	modalTitle = __( 'Add Template Part' ),
```
Comment (youknowriad) [inline packages/edit-site/src/components/create-template-part-modal/index.js]: Can we consistently update all of these use the post type labels?

File: lib/compat/wordpress-6.6/post.php (hunk 2 above)
Comment (ntsekouras) [reply lib/compat/wordpress-6.6/post.php]: >I can extract the labels update to remove the new for other post types too in a separate PR

I'd create a GB PR for that, but you suggest skip it altogether and only do it in core?

Comment (youknowriad) [reply lib/compat/wordpress-6.6/post.php]: The important thing is that it should update all the post types to stay consistent (not just the ones shown in the site editor). I suggested to do it in Core because it impacts the UI of WP-Admin entirely and Core might give more visibility to this change. It shouldn't be hidden in the "Add create pattern button" PR.

Comment (jameskoster) [reply lib/compat/wordpress-6.6/post.php]: For reference I opened a trac ticket about this last week; https://core.trac.wordpress.org/ticket/61219#ticket

Comment (ntsekouras) [reply lib/compat/wordpress-6.6/post.php]: Extracted them and opened a core PR: https://github.com/WordPress/wordpress-develop/pull/6585.

Comment (youknowriad) [reply lib/compat/wordpress-6.6/post.php]: Great :)

File: packages/edit-site/src/components/add-new-pattern/index.js [hunk 4]
```diff
@@ -95,20 +101,21 @@ export default function AddNewPattern() {
 		onClick: () => {
 			patternUploadInputRef.current.click();
 		},
-		title: __( 'Import pattern from JSON' ),
+		title: __( 'Import from JSON' ),
```
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-pattern/index.js]: Is this change required, or should we keep specifying that we can only import patterns (not template parts)

File: packages/edit-site/src/components/page-patterns/header.js [hunk 5]
```diff
@@ -57,39 +58,42 @@ export default function PatternsHeader( {
...
-									category={ patternCategory }
-									onClose={ onClose }
-								/>
-							</MenuGroup>
-						) }
-					</DropdownMenu>
-				) }
+				<HStack expanded={ false }>
+					<AddNewPattern />
+					{ !! patternCategory?.id && (
```
Comment (youknowriad) [inline packages/edit-site/src/components/page-patterns/header.js:68]: I think this menu should move close to the "title" (left of the header) rather than staying on the right no? Anyway, not a blocker.

**Review by youknowriad — APPROVED** [review]
Nice work here 👍

File: packages/edit-site/src/components/add-new-pattern/index.js (hunk 4 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/add-new-pattern/index.js]: That's a good question. Designs suggested to remove this, but would like some confirmation from @jameskoster or @jasmussen

Comment (jameskoster) [reply packages/edit-site/src/components/add-new-pattern/index.js]: Good point. Probably fine to leave the explicit label for now.

**Comment by jameskoster** [comment]
It would be really nice to use DropdownMenu V2 here (https://github.com/WordPress/gutenberg/issues/61094). Fine to do that later/separately if it's not trivial.
